### PR TITLE
[#677] Render Suggested Replies

### LINKED ISSUE #677: Render Suggested Replies
### Summary

The following criteria must be met for each suggested reply to be eligible to be rendered:
- ID
- name
- source type (This is necessary because we have to know for which source they are. We need to make it possible to have some replies that are source agnostic.)
- content (This will be very source specific, so our system should accept whatever JSON is saved because there is no CMS feature on our side.)

### Requirements

If a question is received by a system user on Airy Core, it should be possible to view suggested replies. If a system user clicks the suggested reply modal, we should offer the option of one or several reply options. Once the system user selects a suggested reply to send, it should auto-fill the input bar and the system user should be able to send this message.

02 0 Suggested Replies one option@2x
02 1 Suggested Replies one option@2x
02 2 Suggested Replies one option@2x
01 0 Suggested Replies – many options@2x
01 1 0 Suggested Replies – many options@2x
01 1 1 Suggested Replies – many options@2x
01 2 Suggested Replies@2x
01 3 Suggested Replies@2x

We need to be able to render different types of templates regardless of whether they are composed of text, media, or buttons.

03 0 Suggested Replies – Templates@2x
03 1 Suggested Replies – Templates@2x
03 2 Suggested Replies – Templates@2x

We should be able to render templates composed of long messages. We should be able to display a cropped view of these replies and furthermore, system users should be able to send these messages in full.

04 0 Long message crop@2x

It should be possible for system users to be prompted with more than 1 intent. The system user should be able to choose from these replies and send the message.

05 0 Suggested Replies – more than one intent@2x

### Design Note

When creating this feature, we should only include the template button within the input bar. Do not include the template button next to the suggested replies.

### PR DESCRIPTION
**Pull request by bitboxer**
closes #677

### CODE REVIEW

**Comment by bitboxer** [comment]
This is how it looks:

Screen Shot 2021-03-22 at 15 19 52

File: frontend/ui/src/pages/Inbox/MessageInput/index.tsx [hunk 1]
```diff
@@ -189,54 +196,95 @@ const MessageInput = (props: MessageInputProps & ConnectedProps<typeof connector
...
 
+  const getLastMessageWithSuggestedReplies = useCallback(() => {
+    const lastMessages = props.messages
+      ?.filter((message: Message) => message.senderType == 'source_contact')
+      .slice(props.messages.length - 5)
+      .reverse();
+    return lastMessages?.find(
+      (message: Message) => message.metadata?.suggestions && Object.keys(message.metadata.suggestions).length > 0
+    );
+  }, [props.messages]);
```
Comment (chrismatix) [inline frontend/ui/src/pages/Inbox/MessageInput/index.tsx:207]: So the idea here is that if the last message does not have suggestions we assume that any of the previous 5 message might have suggestions that will also work? :) 

I don't really have an opinion on whether or not that's good. Just want to see if I understand the idea.

File: frontend/ui/src/reducers/data/messages/index.ts [hunk 2]
```diff
@@ -25,6 +26,63 @@ function mergeMessages(oldMessages: Message[], newMessages: Message[]): Message[
   return sortBy(messages, message => message.sentAt);
 }
 
+const findConversationId = (state: Messages, messageId: string) => {
```
Comment (chrismatix) [inline frontend/ui/src/reducers/data/messages/index.ts:29]: Ok this is a funny consequence of the data model I had not anticipated. This means that the update becomes `O(n)` compared to `O(1)` compared to the other stores.

Comment (chrismatix) [reply frontend/ui/src/reducers/data/messages/index.ts:29]: But I don't have a better suggestion other than populating a separate field on the store with a lookup table to amortize the cost

**Review by chrismatix — APPROVED** [review]
Good work!

File: frontend/ui/src/pages/Inbox/MessageInput/index.tsx (hunk 1 above)
Comment (bitboxer) [reply frontend/ui/src/pages/Inbox/MessageInput/index.tsx:207]: Yes, it was outlined in the ticket this way.

File: frontend/ui/src/reducers/data/messages/index.ts (hunk 2 above)
Comment (bitboxer) [reply frontend/ui/src/reducers/data/messages/index.ts:29]: ...or just adding the conversation id in the websocket message. That would make this also a `O(1)` thing again.

**Pull request merged**